sirepo/pkcli/service.py

```diff
@@ -13,63 +13,75 @@
+import errno
+import os
 import py
 import re
+import signal
 import sirepo.srdb
 import socket
+import subprocess
+import sys
 
 
-def celery():
-    """Start celery"""
-    assert pkconfig.channel_in('dev')
-    import celery.bin.celery
-    import sirepo.celery_tasks
-    run_dir = _run_dir().join('celery').ensure(dir=True)
-    with pkio.save_chdir(run_dir):
-        celery.bin.celery.main(argv=[
-            'celery',
-            'worker',
-            '--app=sirepo.celery_tasks',
-            '--no-color',
-            '-Ofair',
-            '--queue=' + ','.join(sirepo.celery_tasks.QUEUE_NAMES),
-        ])
-
-
-def flower():
-    """Start flower"""
-    from flower import command
-    assert pkconfig.channel_in('dev')
-    run_dir = _run_dir().join('flower').ensure(dir=True)
-    with pkio.save_chdir(run_dir):
-        command.FlowerCommand().execute_from_commandline([
-            'flower',
-            '--address=' + cfg.ip,
-            '--app=sirepo.celery_tasks',
-            '--no-color',
-            '--persistent',
-        ])
+def flask():
+    from sirepo import server
+
+    use_reloader = pkconfig.channel_in('dev')
+    app = server.init(use_reloader=use_reloader)
+    # avoid WARNING: Do not use the development server in a production environment.
+    app.env = 'development'
+    app.run(
+        host=cfg.ip,
+        port=cfg.port,
+        threaded=True,
+        use_reloader=use_reloader,
+    )
 
 
 def http():
-    """Starts Flask server in http mode.
+    """Starts the Flask server and job_supervisor.
 
     Used for development only.
     """
-    from sirepo import server
-
-    with pkio.save_chdir(_run_dir()):
-        use_reloader = pkconfig.channel_in('dev')
-        app = server.init(use_reloader=use_reloader)
-        # avoid WARNING: Do not use the development server in a production environment.
-        app.env = 'development'
-        app.run(
-            host=cfg.ip,
-            port=cfg.port,
-            threaded=True,
-            use_reloader=use_reloader,
+    def _env():
+        e = os.environ
+        e.update(
+            SIREPO_JOB_DRIVER_MODULES='local',
         )
+        return e
+
+    def _signal_exit(*args):
+        _exit(*args)
+        sys.exit(1)
+
+    def _exit(*args):
+        for p in processes:
+            try:
+                p.terminate()
+                p.wait(1)
+            except ProcessLookupError:
+                continue
+            except subprocess.TimeoutExpired:
+                p.kill()
+
+    def _start(service):
+        c = ['pyenv', 'exec', 'sirepo']
+        c.extend(service)
+        processes.append(subprocess.Popen(
+            c,
+            cwd=str(_run_dir()),
+            env=_env(),
+        ))
+
+    processes = []
+    signal.signal(signal.SIGINT, _signal_exit)
+    signal.signal(signal.SIGTERM, _signal_exit)
+    _start(['job_supervisor'])
+    _start(['service', 'flask'])
+    p, _ = os.wait()
+    _exit(signal.SIGTERM)
```

Comment:
This is a little bit odd, but you should reset the signal handler after exiting `http` in a try/finally. This is global state, and if it is ever called directly, you'll want to reset it. `pkcli` functions should be directly callable.

The try/finally should also handle killing the processes. This is also global state. If there's some type of unexpected exception, then the try finally should stop the processes. In this case, it should be BaseException, not just Exception, because it is global state.

In thinking about it, the signal handler does not need to sys.exit. If the signal handler sends a kill, the os.wait in `http` will return, and then the normal sending of signals (to dead processes, but that doesn't matter) will proceed and a normal return from `http` can happen.

The os.wait may raise `ChildProcessError` so you'll need to catch that.

Reply:
Makes sense. I believe I resolved everything.

I'm using a try/finally to start/kill the subprocesses. Regarding the bit onBaseException, do I need to worry about it? If I'm killing in a finally then the type of exception doesn't matter. Do you want me to catch BaseException and do a pass to smother any errors? Seems like we'd want those exceptions to bubble up to the caller.

Reply:
You are correct: finally/with is enough.